src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Heaps.cs

```diff
@@ -290,14 +290,14 @@ internal void CompleteHeaps()
 
         public ImmutableArray<int> GetHeapSizes()
         {
-            var heapSizes = new int[MetadataTokens.HeapCount];
+            var builder = ImmutableArray.CreateBuilder<int>(MetadataTokens.HeapCount);
 
-            heapSizes[(int)HeapIndex.UserString] = _userStringWriter.Count;
-            heapSizes[(int)HeapIndex.String] = _stringWriter.Count;
-            heapSizes[(int)HeapIndex.Blob] = _blobHeapSize;
-            heapSizes[(int)HeapIndex.Guid] = _guidWriter.Count;
+            builder[(int)HeapIndex.UserString] = _userStringWriter.Count;
```

Comment:
There must not be tests exercising your changes.  Both this change and the one below in MetadataBuilder.Tables.cs are going to throw exceptions. Please ensure there are tests covering any such changes you make, in addition to fixing this.

(Also, you talk about this saving an allocation.  Note that the builder is itself a class, so this doesn't change the number of allocations, though it does change the size of them, in that it's allocating a builder instead of another array, and save on a copy.)